diffcp/utils.py

```diff
@@ -36,3 +37,169 @@ def get_random_like(A, randomness):
+    subject to  Ax + s = b-tau*dy
+                s \in K
+    This is just a perturbed instance of the forward optimization problem.
+
+    For ds!=0 we rewrite the problem as an embedded problem.
+    We add a constraint x'=s, replace all appearances of s with x' and solve
+    argmin.     <[x, x'], [[P+rho*I,     0];  [x, x']> + <[c+tau*dx-rho*x^*, tau*ds-rho*s^*], [x, x']>
+                           [      0, rho*I]]
+    subject to  [[A,  I];  [x, x'] + [s'; s] = [b-tau*dy; 0]
+                 [0, -I]]
+                (s', s) \in (0 x K)
+    Note that we also add a regularizer on s in this case (rho/2 |s-s^*|^2).
+    
+    Args:
+        dx: NumPy array representing perturbation in `x`
+        dy: NumPy array representing perturbation in `y`
+        ds: NumPy array representing perturbation in `s`
+        tau: Perturbation strength parameter
+        rho: Regularization strength parameter
+    Returns:
+        x_pert: Perturbed solution of the primal variable x
+        y_pert: Perturbed solution of the dual variable y
+        s_pert: Perturbed solution of the slack variable s
+    """
+    m, n = A.shape
+
+    # The cases ds = 0 and ds != 0 are handled separately, see docstring
+    if np.isclose(np.sum(np.abs(ds)), 0):
+        # Perturb problem (Note: perturb primal and dual linear terms in different directions)
+        c_pert = c + tau * dx
+        b_pert = b - tau * dy
+
+        # Regularize: Effectively adds a rho/2 |x-x^*|^2 term to the objective
+        P_reg = regularize_P(P, rho=rho, size=n)
+        c_pert_reg = c_pert - rho * x
+
+        # Set warm start
+        warm_start = (x, y, s) if solve_method != "ECOS" else None
+
+        # Solve the perturbed problem
+        # Note: In special case solve_method=='SCS' and rho==0, this could be sped up strongly by using solver.update
+        result_pert = solve_internal(A=A, b=b_pert, c=c_pert_reg, P=P_reg, cone_dict=cone_dict, 
+                                        solve_method=solve_method, warm_start=warm_start, **solver_kwargs)
+        # Extract the solutions
+        x_pert, y_pert, s_pert = result_pert["x"], result_pert["y"], result_pert["s"]
+    else:
+        # Embed problem (see docstring)
+        A_emb, b_emb, c_emb, P_emb, cone_dict_emb = embed_problem(A, b, c, P, cone_dict)
+
+        # Perturb problem (Note: perturb primal and dual linear terms in different directions)
+        b_emb_pert = b_emb - tau * np.hstack([dy, np.zeros(m)])
+        c_emb_pert = c_emb + tau * np.hstack([dx, ds])
+        
+        # Regularize: Effectively adds a rho/2 (|x-x^*|^2 + |s-s^*|^2) term to the objective
+        P_emb_reg = regularize_P(P_emb, rho=rho, size=n+m)
+        c_emb_pert_reg = c_emb_pert - rho * np.hstack([x, s])
+
+        # Set warm start
+        if solve_method == "ECOS":
+            warm_start = None
+        else:
+            warm_start = (np.hstack([x, s]), np.hstack([y, y]), np.hstack([s, s]))
```

Comment:
Does Clarabel use a warmstart?

Reply:
It does not seem like they support it, currently in diffcp the warm_start is also not passed to the Clarabel solver so it definitely is not used at the moment. Probably better to make this explicit, I will add a check to throw an error when trying to use warmstarteing with Clarabel.